File: eecs268/2020spring/video-lectures/2020.04.15.binarytrees-part-2/BinaryTree.cpp

//public search
template <typename T>
bool BinaryTree<T>::search(T entry)
{
	return(search(entry, m_root));
}

//public count
template <typename T>
bool BinaryTree<T>::search(T entry)
{
	return(count(???));
}

//private search
template <typename T>
bool BinaryTree<T>::search(T entry, BinaryNode<T>* subtree)
{
	//base cases
	//1) if subtree is nullptr then the subtree we're 
	//    searching is empty. Return false
	//2) if we've found it, return true
	//NOTE always check for nullptr first!
	if(subtree == nullptr)
	{
		return(false);
	}
	else if(subtree->getEntry() == entry)
	{
		return(true);
	}
	else
	{
		//search lst and rst
		bool isInLST = search(entry, subtree->getLeft());
		bool isInRST = search(entry, subtree->getRight());
		return( isInLST || isInRST );
	}
}

File: eecs268/2025spring/MWF/2025.03.26.binary-trees/binarytree.py
#binarytree.py

from binarynode import BinaryNode

class BinaryTree:
    def __init__(self):
        self._root = None

    def add(self, entry):
        pass
        #we'll assume this works for now

    def search(self, target):
        return self._rec_search(target, self._root)

    def _rec_search(self, target, cur_node):
        if cur_node is None:
            return False
        elif cur_node.entry == target:
            return True
        else:
            #search the LST and RST
            is_in_LST = self._rec_search(target, cur_node.left)
            is_in_RST = self._rec_search(target, cur_node.right)
            return is_in_LST or is_in_RST

    
